src/transformation/visitors/switch.ts

```diff
@@ -1,61 +1,152 @@
+    const clauses = statement.caseBlock.clauses;
+    if (clauses.length === 1 && ts.isDefaultClause(clauses[0])) {
+        const defaultClause = clauses[0].statements;
+        if (defaultClause.length) {
+            statements.push(lua.createDoStatement(context.transformStatements(defaultClause)));
         }
+    } else {
+        // Build up the condition for each if statement
+        // Fallthrough is handled by accepting the last condition as an additional or clause
+        // Default is the not of all known case expressions
+        let previousClause: ts.CaseOrDefaultClause;
+        let condition: lua.Expression | undefined;
+        statement.caseBlock.clauses.forEach(clause => {
+            if (!condition || (previousClause && containsBreakOrReturn([...previousClause.statements]))) {
+                if (ts.isDefaultClause(clause)) {
+                    // If the default is first, or followed by a break, we can't fall into it, skip.
+                } else {
+                    condition = lua.createBinaryExpression(
+                        switchVariable,
+                        context.transformExpression(clause.expression),
+                        lua.SyntaxKind.EqualityOperator
+                    );
+                }
+            } else {
+                if (ts.isDefaultClause(clause)) {
+                    // use the previous condition for the default clause
+                } else {
+                    condition = lua.createBinaryExpression(
+                        condition,
+                        lua.createBinaryExpression(
+                            switchVariable,
+                            context.transformExpression(clause.expression),
+                            lua.SyntaxKind.EqualityOperator
+                        ),
+                        lua.SyntaxKind.OrOperator
+                    );
+                }
+            }
 
-        // If the clause condition holds, go to the correct label
-        const condition = lua.createBinaryExpression(
-            switchVariable,
-            context.transformExpression(clause.expression),
-            lua.SyntaxKind.EqualityOperator
-        );
+            if (clause.statements.length) {
+                if (!ts.isDefaultClause(clause) && condition) {
+                    statements.push(
+                        lua.createAssignmentStatement(
+                            conditionVariable,
+                            lua.createBinaryExpression(conditionVariable, condition, lua.SyntaxKind.OrOperator)
+                        )
+                    );
+                    condition = undefined;
+                }
 
-        const goto = lua.createGotoStatement(`${switchName}_case_${index}`);
-        return lua.createIfStatement(condition, lua.createBlock([goto]), previousCondition);
-    }, undefined as lua.IfStatement | undefined);
+                statements.push(
```

Comment:
We should skip this on default clauses that aren't fallen into so we don't output an erroneous if block that won't be hit.

Reply:
Agreed, I spotted this one after I got tests passing.